lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,132 @@
+'use strict';
+
+const kLastResolve = Symbol('lastResolve');
+const kLastReject = Symbol('lastReject');
+const kError = Symbol('error');
+const kEnded = Symbol('ended');
+const kLastPromise = Symbol('lastPromise');
+const kHandlePromise = Symbol('handlePromise');
+
+class AsyncIteratorRecord {
+  constructor(value, done) {
+    this.done = done;
+    this.value = value;
+  }
+}
+
+function readAndResolve(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    const data = iter.stream.read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      resolve(new AsyncIteratorRecord(data, false));
+      iter[kLastResolve] = null;
+      iter[kLastReject] = null;
+    }
+  }
+}
+
+function onReadable(iter) {
+  // we wait for the next tick, because it might
+  // emit an error with process.nextTick
+  process.nextTick(readAndResolve, iter);
+}
+
+function onEnd(iter) {
+  const resolve = iter[kLastResolve];
+  if (resolve !== null) {
+    resolve(new AsyncIteratorRecord(null, true));
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter[kEnded] = true;
+}
+
+function onError(iter, err) {
+  const reject = iter[kLastReject];
+  // reject if we are waiting for data in the Promise
+  // returned by next() and store the error
+  if (reject !== null) {
+    reject(err);
+    iter[kLastResolve] = null;
+    iter[kLastReject] = null;
+  }
+  iter.error = err;
+}
+
+class ReadableAsyncIterator {
+  constructor(stream) {
+    this.stream = stream;
+    this[kLastResolve] = null;
+    this[kLastReject] = null;
+    this[kError] = null;
+    this[kEnded] = false;
+    this[kLastPromise] = null;
+
+    stream.on('readable', onReadable.bind(null, this));
+    stream.on('end', onEnd.bind(null, this));
+    stream.on('error', onError.bind(null, this));
+
+    // the function passed to new Promise
+    // is cached so we avoid allocating a new
+    // closure at every run
+    this[kHandlePromise] = (resolve, reject, promise) => {
+      // if we have multiple next() calls
+      // we will wait for the previous Promise to finish
+      const lastPromise = this[kLastPromise];
+      if (promise === lastPromise) {
+        this[kLastPromise] = null;
+      } else if (promise === undefined && lastPromise !== null) {
+        const boundFn =
+          this[kHandlePromise].bind(this, resolve, reject, lastPromise);
+        lastPromise.then(boundFn, boundFn);
```

Comment:
`.finally(boundFn)`?

Reply:
Do we have the right V8 version for that? It only went in at some point in V8 6.3, not sure if it was before we cut our branch or not. Also, I believe it has slightly different semantics in the spec so it should be benchmarked.

Reply:
FWIW, the `Promise.prototype.finally` and async iteration have been shipped in the same day:

https://bugs.chromium.org/p/v8/issues/detail?id=5855#c36
https://bugs.chromium.org/p/v8/issues/detail?id=5967#c11

Reply:
FWIW, the `Promise.prototype.finally` and async iteration have been shipped in the same day:

https://bugs.chromium.org/p/v8/issues/detail?id=5855#c36
https://bugs.chromium.org/p/v8/issues/detail?id=5967#c11